docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+
+Having a mechanism to sync an initial set of default resources after clusters are created makes clusters created with Cluster API functional and ready for workloads from the beginning, without requiring additional user intervention. 
+
+### Goals
+
+Provide a means to specify a set of resources to apply automatically to newly-created and existing Clusters. Resources initially will be applied only once, adding sync functionality is optional
+Support additions to the resource list by applying the new added resources to both new and existing matching clusters
+Support both json and yaml resources
+
+### Non-Goals/Future Work
+
+Replace or compete with the Cluster Addons subproject
+Support deletion of resources. Deleting resources from the list will not result in deletion of those resources from the synced clusters, but deleted resources will not be applied to the new clusters
+Responsible for cleaning up resources applied by ClusterResourceSet controller when ClusterResourceSet resource is deleted
+Responsible for lifecycle management of the installed resources (such as CNI)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce" / "Sync"
```

Comment:
Have folks considered re-using the "EnsureExists" / "Reconcile" nomenclature from the Addon Manager project:

Ref:

https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager

I think this is conceptually equivalent, it might be beneficial to not introduce add'l terms here.

Reply:
Thanks for pointing this out @jackfrancis.
I think for `Sync` can be replaced with `Reconcile` but `EnsureExists` does not actually describe what `ApplyOnce` is doing. 
The idea is simpler here, apply all resources to a cluster only once. We do not worry about if the resources exist or not, because the proposed use case of `ApplyOnce` is to apply resources to newly created clusters. So, we do not reconcile resources if for some reason the resources applied by ClusterResourseSet gets deleted by some other entity; this may fall into lifecycle management territory, which is not the purpose here.

Reply:
That makes sense, thanks for pointing that out. In the kube-addon-manager case, both modes are flavors of reconciliation, where in this case, we only have a single reconciliation mode.

Are we essentially declaring "Create If Not Exists"? I.e., what does `ApplyOnce` do if the resource already exists due to the outcome of some out of band (from the persepctive of cluster-api) operation? Do we simply no-op, or do we throw an error?

Reply:
If the resources exist, we can do CreateOrUpdate() to cover cases like if there is an existing Secret with the same name as the resource, we should still update its data to say that we applied that resource with ClusterResourceSet.

Will keep this thread open for any opinions on this.

Reply:
If a resource already exist I can see that we might end up overwriting user-applied changes if we reconcile it again. I'd lean on (for this first iteration) the safer side, and avoid updating or patching any pre-existing objects.

Reply:
I think not reapplying an existing resource breaks the consistency of ClusterResourceSet resources because a ClusterResourceSet resource may rely on the data in another ClusterResourceSet resource (which pre-exist in the cluster but with different data). By not reapplying, we may end up non-functional resources and still see success in the status ConfigMap that shows which ClusterResourceSets are applied to that particular cluster.

Reply:
That makes sense, do you have a use case I can look at? Sorry if I missed it in the PR

Reply:
Your point of rewriting existing resources is also valid and concerning, not sure which one is more common use case. I made my point considering ApplyOnce is more suitable for fresh installs.
Will update use cases based on the decision in this thread.

Reply:
Yeah, it's a balancing act as everything else, let's get more feedback from users, we might be able to come up with a solution quickly that way?